File: src/main/java/tn/esprit/tahwissa/models/ReservationVoyage.java
package tn.esprit.tahwissa.models;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * ReservationVoyage Model - Maps to 'reservationvoyage' table in database
 */
public class ReservationVoyage {
    private int id;
    private int idUtilisateur;
    private int idVoyage;
    private LocalDateTime dateReservation;
    private StatutReservation statut;
    private int nbrPersonnes;
    private BigDecimal montantTotal;
    private LocalDateTime dateCreation;
    
    // For display purposes - joined data from other tables
    private String nomClient;
    private String destinationVoyage;
    private String titreVoyage;

    // Enum for reservation status
    public enum StatutReservation {
        EN_ATTENTE("EN_ATTENTE", "En Attente"),
        CONFIRMEE("CONFIRMEE", "Confirmée"),
        ANNULEE("ANNULEE", "Annulée"),
        TERMINEE("TERMINEE", "Terminée");

        private final String code;
        private final String label;

        StatutReservation(String code, String label) {
            this.code = code;
            this.label = label;
        }

        public String getCode() {
            return code;
        }

        public String getLabel() {
            return label;
        }

        public static StatutReservation fromCode(String code) {
            for (StatutReservation status : values()) {
                if (status.code.equals(code)) {
                    return status;
                }
            }
            return EN_ATTENTE;
        }
    }

    // Constructors
    public ReservationVoyage() {
    }

    public ReservationVoyage(int id, int idUtilisateur, int idVoyage, 
                            LocalDateTime dateReservation, StatutReservation statut,
                            int nbrPersonnes, BigDecimal montantTotal) {
        this.id = id;
        this.idUtilisateur = idUtilisateur;
        this.idVoyage = idVoyage;
        this.dateReservation = dateReservation;
        this.statut = statut;
        this.nbrPersonnes = nbrPersonnes;
        this.montantTotal = montantTotal;
    }

    // Getters and Setters
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getIdUtilisateur() {
        return idUtilisateur;
    }

    public void setIdUtilisateur(int idUtilisateur) {
        this.idUtilisateur = idUtilisateur;
    }

    public int getIdVoyage() {
        return idVoyage;
    }

    public void setIdVoyage(int idVoyage) {
        this.idVoyage = idVoyage;
    }

    public LocalDateTime getDateReservation() {
        return dateReservation;
    }

    public void setDateReservation(LocalDateTime dateReservation) {
        this.dateReservation = dateReservation;
    }

    public StatutReservation getStatut() {
        return statut;
    }

    public void setStatut(StatutReservation statut) {
        this.statut = statut;
    }

    public int getNbrPersonnes() {
        return nbrPersonnes;
    }

    public void setNbrPersonnes(int nbrPersonnes) {
        this.nbrPersonnes = nbrPersonnes;
    }

    public BigDecimal getMontantTotal() {
        return montantTotal;
    }

    public void setMontantTotal(BigDecimal montantTotal) {
        this.montantTotal = montantTotal;
    }

    public LocalDateTime getDateCreation() {
        return dateCreation;
    }

    public void setDateCreation(LocalDateTime dateCreation) {
        this.dateCreation = dateCreation;
    }

    public String getNomClient() {
        return nomClient;
    }

    public void setNomClient(String nomClient) {
        this.nomClient = nomClient;
    }

    public String getDestinationVoyage() {
        return destinationVoyage;
    }

    public void setDestinationVoyage(String destinationVoyage) {
        this.destinationVoyage = destinationVoyage;
    }

    public String getTitreVoyage() {
        return titreVoyage;
    }

    public void setTitreVoyage(String titreVoyage) {
        this.titreVoyage = titreVoyage;
    }

    @Override
    public String toString() {
        return "ReservationVoyage{" +
                "id=" + id +
                ", idUtilisateur=" + idUtilisateur +
                ", idVoyage=" + idVoyage +
                ", dateReservation=" + dateReservation +
                ", statut=" + statut +
                ", nbrPersonnes=" + nbrPersonnes +
                ", montantTotal=" + montantTotal +
                ", nomClient='" + nomClient + '\'' +
                ", destinationVoyage='" + destinationVoyage + '\'' +
                '}';
    }
}

File: IdeaProjects/tahwissa-symfony/src/Entity/ReservationVoyage.php
<?php

namespace App\Entity;

use App\Repository\ReservationVoyageRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: ReservationVoyageRepository::class)]
#[ORM\Table(name: 'reservationvoyage')]
class ReservationVoyage
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: Types::INTEGER)]
    private ?int $id = null;

    #[ORM\Column(name: 'idUtilisateur', type: Types::INTEGER)]
    #[Assert\NotBlank(message: 'L\'identifiant utilisateur est obligatoire')]
    #[Assert\Positive(message: 'L\'identifiant utilisateur doit être positif')]
    private ?int $idUtilisateur = null;

    #[ORM\ManyToOne(targetEntity: Voyage::class, inversedBy: 'reservations')]
    #[ORM\JoinColumn(name: 'id_voyage', referencedColumnName: 'id', nullable: false)]
    #[Assert\NotNull(message: 'Le voyage est obligatoire')]
    private ?Voyage $voyage = null;

    #[ORM\Column(name: 'date_reservation', type: Types::DATETIME_MUTABLE, nullable: true)]
    #[Assert\NotBlank(message: 'La date de réservation est obligatoire')]
    private ?\DateTimeInterface $dateReservation = null;

    #[ORM\Column(type: Types::STRING, length: 20, enumType: null)]
    #[Assert\NotBlank(message: 'Le statut est obligatoire')]
    #[Assert\Choice(
        choices: ['EN_ATTENTE', 'CONFIRMEE', 'ANNULEE', 'TERMINEE'],
        message: 'Le statut doit être EN_ATTENTE, CONFIRMEE, ANNULEE ou TERMINEE'
    )]
    private ?string $statut = 'EN_ATTENTE';

    #[ORM\Column(name: 'nbrPersonnes', type: Types::INTEGER)]
    #[Assert\NotBlank(message: 'Le nombre de personnes est obligatoire')]
    #[Assert\Positive(message: 'Le nombre de personnes doit être au moins 1')]
    #[Assert\LessThanOrEqual(value: 50, message: 'Le nombre de personnes ne peut pas dépasser 50')]
    private ?int $nbrPersonnes = 1;

    #[ORM\Column(name: 'montantTotal', type: Types::DECIMAL, precision: 10, scale: 2)]
    private ?string $montantTotal = '0.00';

    #[ORM\Column(name: 'dateCreation', type: Types::DATETIME_MUTABLE)]
    private ?\DateTimeInterface $dateCreation = null;

    public function __construct()
    {
        $this->dateCreation = new \DateTime();
        $this->dateReservation = new \DateTime();
        $this->statut = 'EN_ATTENTE';
        $this->nbrPersonnes = 1;
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getIdUtilisateur(): ?int
    {
        return $this->idUtilisateur;
    }

    public function setIdUtilisateur(int $idUtilisateur): static
    {
        $this->idUtilisateur = $idUtilisateur;
        return $this;
    }

    public function getVoyage(): ?Voyage
    {
        return $this->voyage;
    }

    public function setVoyage(?Voyage $voyage): static
    {
        $this->voyage = $voyage;
        return $this;
    }

    public function getDateReservation(): ?\DateTimeInterface
    {
        return $this->dateReservation;
    }

    public function setDateReservation(?\DateTimeInterface $dateReservation): static
    {
        $this->dateReservation = $dateReservation;
        return $this;
    }

    public function getStatut(): ?string
    {
        return $this->statut;
    }

    public function setStatut(string $statut): static
    {
        $this->statut = $statut;
        return $this;
    }

    public function getNbrPersonnes(): ?int
    {
        return $this->nbrPersonnes;
    }

    public function setNbrPersonnes(int $nbrPersonnes): static
    {
        $this->nbrPersonnes = $nbrPersonnes;
        return $this;
    }

    public function getMontantTotal(): ?string
    {
        return $this->montantTotal;
    }

    public function setMontantTotal(string $montantTotal): static
    {
        $this->montantTotal = $montantTotal;
        return $this;
    }

    public function getDateCreation(): ?\DateTimeInterface
    {
        return $this->dateCreation;
    }

    public function setDateCreation(\DateTimeInterface $dateCreation): static
    {
        $this->dateCreation = $dateCreation;
        return $this;
    }

    /**
     * Calculates and sets the total amount based on voyage price and number of persons
     */
    public function calculerMontantTotal(): static
    {
        if ($this->voyage && $this->nbrPersonnes) {
            $prix = (float) $this->voyage->getPrixUnitaire();
            $this->montantTotal = number_format($prix * $this->nbrPersonnes, 2, '.', '');
        }
        return $this;
    }

    public function getStatutLabel(): string
    {
        return match ($this->statut) {
            'EN_ATTENTE' => 'En attente',
            'CONFIRMEE' => 'Confirmée',
            'ANNULEE' => 'Annulée',
            'TERMINEE' => 'Terminée',
            default => $this->statut,
        };
    }

    public function getStatutBadgeClass(): string
    {
        return match ($this->statut) {
            'EN_ATTENTE' => 'warning',
            'CONFIRMEE' => 'success',
            'ANNULEE' => 'danger',
            'TERMINEE' => 'info',
            default => 'secondary',
        };
    }
}
